decode.go

```diff
@@ -561,6 +574,35 @@ func (dm *decMode) Unmarshal(data []byte, v interface{}) error {
 	return d.value(v)
 }
 
+// UnmarshalPrefix parses the first CBOR-encoded item in data into the value
+// pointed to by v using dm decoding mode.  If v is nil, not a pointer, or a nil pointer,
+// UnmarshalPrefix returns an error. Any remaining bytes after a single valid item
+// has been parsed will be returned in rest.
+//
+// See the documentation for UnmarshalPrefix for details.
+func (dm *decMode) UnmarshalPrefix(data []byte, v interface{}) (rest []byte, err error) {
+	d := decoder{data: data, dm: dm}
+
+	// check valid
+	off := d.off        // Save offset before data validation
+	err = d.valid(true) // allow extra data after valid data item
+	d.off = off         // Restore offset
+
+	// If it is valid, parse the value. This is structured like this to allow
+	// better test coverage
+	if err == nil {
+		err = d.value(v)
+	}
+
+	// If either valid or value returned an error, do not return rest bytes
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the rest of the data slice (which might be len 0)
+	return data[d.off:], nil
```

Comment:
nit: `data` works in this case.  However, since `d.off` is the offset for `d.data`, it is safer to use `d.data` in case it gets out of sync with `data`.

```suggestion
	return d.data[d.off:], nil
```

Reply:
Changed